File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        d = deque()
        d.append(0)
        ans = []
        for i in range(1, k):
            while d and nums[i] > nums[d[-1]]:
                d.pop()
            d.append(i)
        ans.append(nums[d[0]])
        for i in range(k, len(nums)):
            # and element (i - k) will always expire
            # just remove it from deque if it's present
            if d[0] == i - k:
                d.popleft()
            while d and nums[i] > nums[d[-1]]:
                d.pop()
            d.append(i)
            ans.append(nums[d[0]])
        return ans

File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> ans;
        ans.reserve(n - k + 1);

        // Pre-allocate a flat array as our deque
        vector<int> q(n);
        int head = 0, tail = 0; // head..tail is the deque range

        for (int i = 0; i < n; ++i) {
            // Remove out-of-bound indices
            if (head < tail && q[head] <= i - k) {
                head++;
            }
            // Maintain monotonic order
            while (head < tail && nums[q[tail - 1]] < nums[i]) {
                tail--;
            }
            q[tail++] = i;

            if (i >= k - 1) {
                ans.push_back(nums[q[head]]);
            }
        }
        return ans;
    }
};